### PR TITLE
feat(design): add paginator component

### PR DESCRIPTION
**Pull request by lderrickable**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the new behavior?
Add the paginator component to `@daffodil/design` for controlling selection of pages of items.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by xelaint** [comment]
All the spec files not related to paginator should be moved to a separate PR?

File: libs/design/src/molecules/paginator/paginator.component.html [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<span (click)="onNotifyPrevPageChange()" *ngIf="showPrevPageChange()" class="daff-paginator__prev daff-paginator__element">Prev</span>
```
Comment (xelaint) [inline libs/design/src/molecules/paginator/paginator.component.html]: `previous` and `next` should have its own style, independent of `element/page-item`

```
display: flex;
margin-right: 20px; / margin-left: 20px; for next
align-items: center;
```

File: libs/design/src/molecules/paginator/paginator.component.html [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+<span class="daff-paginator__elipsis" *ngIf="showFirstElipsis()"><span class="daff-paginator__element" (click)="onNotifyPageChange(1)" [class.daff-paginator__element--selected]="isSelected(1)">1</span>...</span>
```
Comment (xelaint) [inline libs/design/src/molecules/paginator/paginator.component.html]: `daff-paginator__page-link` instead of `daff-paginator__element`. And does this need to be an `<a>` instead of `<span>`? If it does, it'll affect styling.

File: libs/design/src/molecules/paginator/paginator.component.html [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+<span class="daff-paginator__element-wrapper" *ngFor="let pageNumber of _numberOfPagesArray">
+  <span 
+    class="daff-paginator__element" 
+    [class.daff-paginator__element--selected]="isSelected(pageNumber + 1)" 
+    *ngIf="showNumber(pageNumber + 1)"
+    (click)="onNotifyPageChange(pageNumber + 1)"  
+  >{{ pageNumber + 1 }}</span>
+</span>
+<span class="daff-paginator__elipsis" *ngIf="showLastElipsis()">...<span class="daff-paginator__element" (click)="onNotifyPageChange(_numberOfPages)" [class.daff-paginator__element--selected]="isSelected(_numberOfPages)">{{ _numberOfPages }}</span></span>
```
Comment (xelaint) [inline libs/design/src/molecules/paginator/paginator.component.html]: `ellipsis`

File: libs/design/src/molecules/paginator/paginator.component.html [hunk 4]
```diff
@@ -0,0 +1,14 @@
+<span (click)="onNotifyPrevPageChange()" *ngIf="showPrevPageChange()" class="daff-paginator__prev daff-paginator__element">Prev</span>
+
+<span class="daff-paginator__elipsis" *ngIf="showFirstElipsis()"><span class="daff-paginator__element" (click)="onNotifyPageChange(1)" [class.daff-paginator__element--selected]="isSelected(1)">1</span>...</span>
+<span class="daff-paginator__element-wrapper" *ngFor="let pageNumber of _numberOfPagesArray">
```
Comment (xelaint) [inline libs/design/src/molecules/paginator/paginator.component.html]: `daff-paginator__page-item` instead of `daff-paginator__element-wrapper`

File: libs/design/src/molecules/paginator/paginator.component.scss [hunk 5]
```diff
@@ -0,0 +1,42 @@
+@import '../../scss/daff-util';
+
+:host {
+	display: flex;
+	justify-content: center;
+
+	.daff-paginator {
```
Comment (xelaint) [inline libs/design/src/molecules/paginator/paginator.component.scss]: `.daff-paginator` doesn't need to be inside of `:host`?

File: libs/design/src/molecules/paginator/paginator.component.scss [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+		&__element {
+			@include clickable();
+			padding: 5px 10px;
+
+			&--selected {
```
Comment (xelaint) [inline libs/design/src/molecules/paginator/paginator.component.scss]: Moved up to `&__element/&__page-link`

File: libs/design/src/molecules/paginator/paginator.component.scss [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+		&__element-wrapper {
+			padding: 5px 0;
+		}
+
+		&__elipsis {
+			margin: 5px 0;
+		}
+
+		&__prev {
```
Comment (xelaint) [inline libs/design/src/molecules/paginator/paginator.component.scss]: Use FontAwesome

File: libs/design/src/molecules/paginator/paginator.component.scss [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+			&:before {
+				@include chevron($direction: 'left');
+				margin-right: 5px;
+				font-size: 1.2em;
+			}
+		}
+
+		&__next {
```
Comment (xelaint) [inline libs/design/src/molecules/paginator/paginator.component.scss]: Use FontAwesome

File: libs/design/src/molecules/paginator/paginator.component.scss [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+		&__element {
+			@include clickable();
+			padding: 5px 10px;
+
+			&--selected {
+				border-radius: 3px;
+			}
+		}
+
+		&__element-wrapper {
```
Comment (xelaint) [inline libs/design/src/molecules/paginator/paginator.component.scss]: Why is this needed?

**Comment by lderrickable** [comment]
> All the test files not related to paginator should be moved to a separate PR?

I removed these changes

**Comment by xelaint** [comment]
Also for accessibility, I think paginator needs to be a `<nav>` and use `<ul>` instead of `<span>`s.

https://www.a11ymatters.com/pattern/pagination/

File: libs/design/src/molecules/paginator/paginator.component.scss [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+			padding: 5px 0;
+		}
+
+		&__elipsis {
+			margin: 5px 0;
+		}
+
+		&__prev {
+			&:before {
+				@include chevron($direction: 'left');
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.scss]: done

File: libs/design/src/molecules/paginator/paginator.component.scss [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+				margin-right: 5px;
+				font-size: 1.2em;
+			}
+		}
+
+		&__next {
+
+			&:after {
+				@include chevron($direction: 'right');
```
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.scss]: done

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: fixed

[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: fixed

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 1 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: done

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 2 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: Sure that's fine. Done

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 3 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: fixed

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 4 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: I just removed this class, because it wasn't needed

File: libs/design/src/molecules/paginator/paginator.component.scss (hunk 5 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.scss]: moved

File: libs/design/src/molecules/paginator/paginator.component.scss (hunk 6 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.scss]: It's already under page-link

File: libs/design/src/molecules/paginator/paginator.component.scss (hunk 9 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.scss]: it isn't. removed

File: libs/design/src/molecules/paginator/paginator.component.scss [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+	display: flex;
+	justify-content: center;
+}
+
+.daff-paginator {
+	&__page-link {
+		@include clickable();
+		padding: 5px 10px;
+
+		&--selected {
```
Comment (xelaint) [inline libs/design/src/molecules/paginator/paginator.component.scss]: `border-radius: 3px;` should be moved up to `&__page-link`.

File: libs/design/src/molecules/paginator/paginator.component.scss (hunk 6 above)
Comment (xelaint) [reply libs/design/src/molecules/paginator/paginator.component.scss]: I mean the `border-radius` style. It should be default on the link, not only when it's selected.

**Comment by lderrickable** [comment]
> @lderrickable Also for accessibility, the `page-link`s should use `<li>`s in an unordered list, and `role="navigation"` assigned via HostBinding.
> 
> Checkout this article: https://www.a11ymatters.com/pattern/pagination/

Paginator is now accessible

File: libs/design/src/molecules/paginator/paginator.component.scss (hunk 6 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.scss]: done

File: libs/design/src/molecules/paginator/paginator.component.html [hunk 13]
```diff
@@ -0,0 +1,21 @@
+<nav role="navigation" attr.aria-label="Pagination Navigation">
```
Comment (damienwebdev) [inline libs/design/src/molecules/paginator/paginator.component.html]: The `nav` can be removed since `daff-paginator` can hold the `role`.

File: libs/design/src/molecules/paginator/paginator.component.html [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    <fa-icon [icon]="['fa', 'chevron-left']"></fa-icon>Prev
+  </a>
+
+  <span class="daff-paginator__ellipsis" *ngIf="showFirstEllipsis()"><a tabindex="0" attr.aria-label="Go to Page 1" class="daff-paginator__page-link" (click)="onNotifyPageChange(1)" [class.daff-paginator__page-link--selected]="isSelected(1)">1</a>...</span>
+  <span *ngFor="let pageNumber of _numberOfPagesArray">
+    <a tabindex="0" attr.aria-label="Go to Page {{pageNumber + 1}}"
+      aria-current="isSelected(pageNumber + 1)"
+      class="daff-paginator__page-link" 
+      [class.daff-paginator__page-link--selected]="isSelected(pageNumber + 1)" 
+      *ngIf="showNumber(pageNumber + 1)"
```
Comment (damienwebdev) [inline libs/design/src/molecules/paginator/paginator.component.html]: Could this conditional be built into the calculation of `_numberOfPagesArray` instead?

File: libs/design/src/molecules/paginator/paginator.component.html [hunk 15]
```diff
@@ -0,0 +1,21 @@
+<nav role="navigation" attr.aria-label="Pagination Navigation">
+  <a attr.aria-label="Go to Previous Page" tabindex="0" (click)="onNotifyPrevPageChange()" *ngIf="showPrevPageChange()" class="daff-paginator__prev">
+    <fa-icon [icon]="['fa', 'chevron-left']"></fa-icon>Prev
+  </a>
+
+  <span class="daff-paginator__ellipsis" *ngIf="showFirstEllipsis()"><a tabindex="0" attr.aria-label="Go to Page 1" class="daff-paginator__page-link" (click)="onNotifyPageChange(1)" [class.daff-paginator__page-link--selected]="isSelected(1)">1</a>...</span>
+  <span *ngFor="let pageNumber of _numberOfPagesArray">
+    <a tabindex="0" attr.aria-label="Go to Page {{pageNumber + 1}}"
```
Comment (damienwebdev) [inline libs/design/src/molecules/paginator/paginator.component.html]: Why are all of these `+1`s littered about? Should this calculation be done in the template?

File: libs/design/src/molecules/paginator/paginator.component.html [hunk 16]
```diff
@@ -0,0 +1,21 @@
+<nav role="navigation" attr.aria-label="Pagination Navigation">
+  <a attr.aria-label="Go to Previous Page" tabindex="0" (click)="onNotifyPrevPageChange()" *ngIf="showPrevPageChange()" class="daff-paginator__prev">
+    <fa-icon [icon]="['fa', 'chevron-left']"></fa-icon>Prev
+  </a>
+
+  <span class="daff-paginator__ellipsis" *ngIf="showFirstEllipsis()"><a tabindex="0" attr.aria-label="Go to Page 1" class="daff-paginator__page-link" (click)="onNotifyPageChange(1)" [class.daff-paginator__page-link--selected]="isSelected(1)">1</a>...</span>
```
Comment (damienwebdev) [inline libs/design/src/molecules/paginator/paginator.component.html]: Why is the first element inside the ellipsis?

File: libs/design/src/molecules/paginator/paginator.component.html [hunk 17]
```diff
@@ -0,0 +1,21 @@
+<nav role="navigation" attr.aria-label="Pagination Navigation">
+  <a attr.aria-label="Go to Previous Page" tabindex="0" (click)="onNotifyPrevPageChange()" *ngIf="showPrevPageChange()" class="daff-paginator__prev">
+    <fa-icon [icon]="['fa', 'chevron-left']"></fa-icon>Prev
```
Comment (damienwebdev) [inline libs/design/src/molecules/paginator/paginator.component.html]: Do we need a `peerDependency` on `@fortawesome/font-awesome-angular`?

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 15 above)
Comment (damienwebdev) [inline libs/design/src/molecules/paginator/paginator.component.html]: Considering that these aren't links, can we use `button` instead?

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 16 above)
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: Does it matter? Aren't the labels for communicating functionality? So if they do the same thing, it shouldn't matter that they're labeled the same

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 14 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: I don't think so, because the `*ngFor` iterates over an array and doesn't care what's inside. So we'd need to make `*ngFor` base 1 instead of base 0.

Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: Oh, wait. Actually maybe this can be done

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 13 above)
[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: done

[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: moved

Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: done

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 14 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: done

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 15 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: yep fixed

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 16 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: I changed how the ngFor shows pageNumbers and such, so there's some simplifications I could make with numbers and ellipses. Now the first element isn't inside the ellipsis

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 17 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: Added the peerDependency

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 15 above)
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: changed to buttons

**Comment by lderrickable** [comment]
Also, I made the next and prev disable instead of disappearing.

File: libs/design/src/molecules/paginator/paginator.component.html (hunk 16 above)
[the remark this answers is not in the document]
Comment (damienwebdev) [reply libs/design/src/molecules/paginator/paginator.component.html]: The labels are used for communicating functionality, but they aren't "scoped" to a particular role. So, for example, if there are multiple paginators for different lists, its difficult screen-reader users to tell what paginator item goes with what specific list.

[the remark this answers is not in the document]
Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.html]: done

File: libs/design/src/molecules/paginator/paginator.component.ts [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+@Component({
+  selector: 'daff-paginator',
+  styleUrls: ['./paginator.component.scss'],
+  templateUrl: './paginator.component.html'
+})
+export class DaffPaginatorComponent extends _daffPaginatorBase implements OnInit, DaffColorable {
+  @HostBinding('class.daff-paginator') class = true;
+  @HostBinding('attr.role') role = 'navigation';
+  @Input() color: DaffPalette;
+  @Input() id: string;
```
Comment (damienwebdev) [inline libs/design/src/molecules/paginator/paginator.component.ts]: I think instead we should grab the value of `aria-label` off the host-binding if it exists.

Comment (lderrickable) [reply libs/design/src/molecules/paginator/paginator.component.ts]: done
